modules/images/image-loading-optimization/class-ilo-html-tag-processor.php

```diff
@@ -0,0 +1,296 @@
+		 *
+		 * <html>
+		 *   <head>
+		 *   </head>
+		 *   <body>
+		 *     <p>Hello!</p>
+		 *     <img src="lcp.png">
+		 *   </body>
+		 * </html>
+		 *
+		 * The two upon processing the IMG element, the two arrays should be equal to the following:
+		 *
+		 * $open_stack_tags    = array( 'HTML', 'BODY', 'IMG' );
+		 * $open_stack_indices = array( 0, 1, 1 );
+		 */
+		$this->open_stack_tags    = array();
+		$this->open_stack_indices = array();
+		while ( $p->next_tag( array( 'tag_closers' => 'visit' ) ) ) {
+			$tag_name = $p->get_tag();
+			if ( ! $p->is_tag_closer() ) {
+
+				// Close an open P tag when a P-closing tag is encountered.
+				// TODO: There are quite a few more cases of optional closing tags: https://html.spec.whatwg.org/multipage/syntax.html#optional-tags
+				// Nevertheless, given WordPress's legacy of XHTML compatibility, the lack of closing tags may not be common enough to warrant worrying about any of them.
+				if ( in_array( $tag_name, self::P_CLOSING_TAGS, true ) ) {
+					$i = array_search( 'P', $this->open_stack_tags, true );
+					if ( false !== $i ) {
+						array_splice( $this->open_stack_tags, $i );
+						array_splice( $this->open_stack_indices, count( $this->open_stack_tags ) );
+					}
+				}
+
+				$level                   = count( $this->open_stack_tags );
+				$this->open_stack_tags[] = $tag_name;
+
+				if ( ! isset( $this->open_stack_indices[ $level ] ) ) {
+					$this->open_stack_indices[ $level ] = 0;
+				} else {
+					++$this->open_stack_indices[ $level ];
+				}
+
+				// TODO: We should consider not collecting metrics when the admin bar is shown and the user is logged-in.
+				// Only increment the tag index at this level only if it isn't the admin bar, since the presence of the
+				// admin bar can throw off the indices.
+				if ( 'DIV' === $tag_name && $p->get_attribute( 'id' ) === 'wpadminbar' ) {
+					--$this->open_stack_indices[ $level ];
+				}
+
+				// Now that the breadcrumbs are constructed, yield the tag name so that they can be queried if desired.
+				// Other mutations may be performed to the open tag's attributes by the callee at this point as well.
+				yield $tag_name;
+
+				// Immediately pop off self-closing tags.
+				if ( in_array( $tag_name, self::SELF_CLOSING_TAGS, true ) ) {
+					array_pop( $this->open_stack_tags );
+				}
+			} else {
+				// If the closing tag is for self-closing tag, we ignore it since it was already handled above.
+				if ( in_array( $tag_name, self::SELF_CLOSING_TAGS, true ) ) {
+					continue;
+				}
+
+				// Since SVG and MathML can have a lot more self-closing/empty tags, potentially pop off the stack until getting to the open tag.
+				$did_splice = false;
+				if ( 'SVG' === $tag_name || 'MATH' === $tag_name ) {
+					$i = array_search( $tag_name, $this->open_stack_tags, true );
+					if ( false !== $i ) {
+						array_splice( $this->open_stack_tags, $i );
+						$did_splice = true;
```

Comment:
if encountering an element which _isn't_ an HTML _Void_ element, then regardless of the tag, if the self-closing flag is present, then it closes the tag. all self-closing flags on HTML elements are invalid and ignored, but all on HTML _foreign elements_ are authoritative and obeyed.

```html
<p>There are no empty <div/> tags. This is inside the DIV</p>.
<wp-group>Is a custom element, and so <wp-group /> is self-closing while others aren't.</wp-group>
```

```php
if (
	WP_HTML_Processor::is_void( $tag_name ) || 
	(
		! WP_HTML_Processor::is_html_element( $tag_name ) &&
		$p->has_self_closing_flag
	)
) {
	// this tag immediately closes as soon as we jump to the next tag.
}
```

this requires having a list of all HTML elements, which the HTML Processor currently doesn't do because it doesn't support any, but we'll have to add it. this _can_ lead to common parsing failures because the invalid self-closing flag has become more popular post-React where it's valid and normative in JSX.

Reply:
Very interesting. I wasn't aware of the `WP_HTML_Processor::has_self_closing_flag()` method. 

Nevertheless, maybe having a list of all HTML elements isn't needed here because: (1) if `math` or `svg` are ancestors, we can assume that all tags with self-closing flags will close the tag, and (2) custom elements always have hyphens in them, so if present we can also honor the self-closing tag.

Nevertheless, I just checked your example and it doesn't seem the second example with a custom element is actually true. I adapted your example to use a `span` instead of `div` (since a `div` closes an open `p`):

```html
<p>There are no empty <span/> tags. This is inside the SPAN</p>.
<wp-group>Is a custom element, and so <wp-group /> is self-closing while others aren't.</wp-group>
```

And Chrome renders this as:

![image](https://github.com/WordPress/performance/assets/134745/cf461a7b-3872-4447-9dbd-759e00da9a26)

So perhaps such self-closing foreign elements are limited to MathML and SVG contexts?

Reply:
For example:

```html
<svg>Is a custom element, and so <g /> is self-closing while others aren't.</svg>
```

![image](https://github.com/WordPress/performance/assets/134745/96869f54-436c-409b-bf30-b15e9545045d)

Reply:
by golly, you're right. I have been misinterpreting "foreign element" for a long time now! thankfully I haven't gotten to the place in the HTML Processor where that matters.

thank you very much for pointing this out.